File: delicious-app/lib/features/products/domain/entities/product.dart
// lib/features/products/domain/entities/product.dart
import 'package:equatable/equatable.dart';

class Product extends Equatable {
  final String       id;
  final String       name;
  final String?      description;
  final String?      brand;
  final double       price;
  final double       discountPercent;
  final String       categoryId;
  final List<String> gradientColors;   // always 2 hex strings, e.g. ["#CC0000","#FF4444"]
  final List<String> mood;             // e.g. ["cold","sweet"]
  final String       imageUrl;
  final String?      thumbnailUrl;
  final List<String> imageGallery;
  final List<ProductVariant> variants;
  final ProductNutrition?    nutrition;
  final bool         isAvailable;
  final int          stock;
  final int          preparationTimeMin;
  final bool         isFeatured;
  final ProductRating rating;
  final List<String> tags;

  const Product({
    required this.id,
    required this.name,
    this.description,
    this.brand,
    required this.price,
    this.discountPercent = 0,
    required this.categoryId,
    required this.gradientColors,
    this.mood            = const [],
    required this.imageUrl,
    this.thumbnailUrl,
    this.imageGallery    = const [],
    this.variants        = const [],
    this.nutrition,
    this.isAvailable     = true,
    this.stock           = 0,
    this.preparationTimeMin = 10,
    this.isFeatured      = false,
    this.rating          = const ProductRating(),
    this.tags            = const [],
  });

  double get finalPrice =>
      double.parse((price * (1 - discountPercent / 100)).toStringAsFixed(2));

  bool get hasDiscount   => discountPercent > 0;
  bool get isInStock     => isAvailable && stock > 0;

  /// Primary gradient colour — used for the animated background tween.
  String get primaryColor   => gradientColors.isNotEmpty ? gradientColors[0] : '#FF6B35';
  String get secondaryColor => gradientColors.length > 1  ? gradientColors[1] : '#FF8C61';

  @override
  List<Object?> get props => [id, name, price, categoryId, isAvailable];
}

// ── Value objects ─────────────────────────────────────────────────────────────

class ProductVariant extends Equatable {
  final String id;
  final String label;
  final double price;
  final int    stock;
  const ProductVariant({
    required this.id,
    required this.label,
    required this.price,
    required this.stock,
  });
  @override
  List<Object> get props => [id, label, price];
}

class ProductRating extends Equatable {
  final double average;
  final int    count;
  const ProductRating({this.average = 0, this.count = 0});
  @override
  List<Object> get props => [average, count];
}

class ProductNutrition extends Equatable {
  final int?    calories;
  final double? protein;
  final double? carbs;
  final double? fat;
  final double? sugar;
  final String? servingSize;
  const ProductNutrition({
    this.calories, this.protein, this.carbs,
    this.fat, this.sugar, this.servingSize,
  });
  @override
  List<Object?> get props => [calories, protein, carbs, fat, sugar, servingSize];
}

File: delicious-app/lib/backend/src/models/Product.js
// src/models/Product.js
const mongoose = require('mongoose');

// ── Mood tags — drive the "mood-based ordering" feature ───────────────────────
const MOOD_TAGS = ['hot', 'cold', 'sweet', 'salty', 'spicy', 'fresh', 'energising', 'comforting'];

// ── Embedded: nutritional info (optional, shown on product detail page) ───────
const NutritionSchema = new mongoose.Schema(
  {
    calories:     { type: Number },
    protein:      { type: Number },  // grams
    carbs:        { type: Number },  // grams
    fat:          { type: Number },  // grams
    sugar:        { type: Number },  // grams
    servingSize:  { type: String },  // e.g. "330ml", "250g"
  },
  { _id: false }
);

// ── Embedded: product variant (size / flavour) ────────────────────────────────
const VariantSchema = new mongoose.Schema(
  {
    label:  { type: String, required: true },  // "Small", "Medium", "Large"
    price:  { type: Number, required: true },  // DZD
    stock:  { type: Number, default: 0 },
  },
  { _id: true }
);

// ── Main schema ───────────────────────────────────────────────────────────────

const ProductSchema = new mongoose.Schema(
  {
    // ── Core identity ─────────────────────────────────────────────────────────
    name: {
      type: String,
      required: [true, 'Product name is required'],
      trim: true,
      maxlength: [120, 'Name cannot exceed 120 characters'],
    },
    description: {
      type: String,
      trim: true,
      maxlength: [800, 'Description cannot exceed 800 characters'],
    },
    brand: {
      type: String,
      trim: true,
      // e.g. "Coca-Cola", "Pepsi", "Ifri", null for homemade dishes
    },

    // ── Pricing ───────────────────────────────────────────────────────────────
    price: {
      type: Number,
      required: [true, 'Price is required'],
      min: [0, 'Price cannot be negative'],
    },
    discountPercent: {
      type: Number,
      default: 0,
      min: 0,
      max: 100,
    },
    // Virtual: finalPrice = price * (1 - discountPercent/100)

    // ── Categorisation ────────────────────────────────────────────────────────
    categoryId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Category',
      required: [true, 'Category is required'],
    },

    // ── UI / theming ──────────────────────────────────────────────────────────
    // The two gradient stop colours shown in the product carousel and background.
    // e.g. Coca-Cola: ["#CC0000","#FF4444"], Pepsi: ["#004B93","#0070CC"]
    gradientColors: {
      type: [String],
      required: [true, 'gradientColors is required'],
      validate: {
        validator: (v) => v.length === 2 && v.every(c => /^#[0-9A-Fa-f]{6}$/.test(c)),
        message: 'gradientColors must be exactly 2 valid hex colour strings, e.g. ["#FF0000","#FF8888"]',
      },
    },

    // ── Mood system ───────────────────────────────────────────────────────────
    // Tags that power "I want something cold/sweet/hot" filtering.
    mood: {
      type: [String],
      enum: {
        values: MOOD_TAGS,
        message: `Mood must be one of: ${MOOD_TAGS.join(', ')}`,
      },
      default: [],
    },

    // ── Media ─────────────────────────────────────────────────────────────────
    imageUrl: {
      type: String,
      required: [true, 'Product image is required'],
    },
    thumbnailUrl: { type: String },  // smaller version for carousel
    imageGallery: { type: [String], default: [] },

    // ── Variants (size / flavour options) ────────────────────────────────────
    variants: { type: [VariantSchema], default: [] },

    // ── Nutrition ─────────────────────────────────────────────────────────────
    nutrition: { type: NutritionSchema },

    // ── Availability & stock ──────────────────────────────────────────────────
    isAvailable: { type: Boolean, default: true },
    stock: {
      type: Number,
      default: 0,
      min: [0, 'Stock cannot be negative'],
    },
    preparationTimeMin: {
      type: Number,
      default: 10,  // minutes
    },

    // ── Metadata ──────────────────────────────────────────────────────────────
    isFeatured: { type: Boolean, default: false },
    rating: {
      average: { type: Number, default: 0, min: 0, max: 5 },
      count:   { type: Number, default: 0 },
    },
    tags: { type: [String], default: [] }, // free-form search tags
  },
  {
    timestamps: true,
    toJSON:   { virtuals: true },
    toObject: { virtuals: true },
  }
);

// ── Virtuals ──────────────────────────────────────────────────────────────────

ProductSchema.virtual('finalPrice').get(function () {
  return +(this.price * (1 - this.discountPercent / 100)).toFixed(2);
});

ProductSchema.virtual('hasDiscount').get(function () {
  return this.discountPercent > 0;
});

// ── Indexes ───────────────────────────────────────────────────────────────────
ProductSchema.index({ categoryId: 1 });
ProductSchema.index({ mood: 1 });
ProductSchema.index({ isAvailable: 1 });
ProductSchema.index({ isFeatured: 1 });
ProductSchema.index({ name: 'text', description: 'text', brand: 'text', tags: 'text' }); // full-text search

module.exports = mongoose.model('Product', ProductSchema);